File: Fraction_To_Recurring_Decimal.cpp
/*166. Fraction to Recurring Decimal  QuestionEditorial Solution  My Submissions
Total Accepted: 39665 Total Submissions: 242506 Difficulty: Medium
Given two integers representing the numerator and denominator of a fraction, return the fraction in string format.

If the fractional part is repeating, enclose the repeating part in parentheses.

For example,

Given numerator = 1, denominator = 2, return "0.5".
Given numerator = 2, denominator = 1, return "2".
Given numerator = 2, denominator = 3, return "0.(6)".
Hint:

No scary math, just apply elementary math knowledge. Still remember how to perform a long division?
Try a long division on 4/9, the repeating part is obvious. Now try 4/333. Do you see a pattern?
Be wary of edge cases! List out as many test cases as you can think of and test your code thoroughly.
Credits:
Special thanks to @Shangrila for adding this problem and creating all test cases.

Hide Company Tags Google
Hide Tags Hash Table Math
*/



/*
numerator 分子
denominator 分母
1. check the return is negative or positive, using XOR^ operator
	then change all of them to positive, store them into long, since int might over limit
2. get the integral part
3. get the fractional part and store them into HashMap<Long, Integer>
	each time num*10 to get the next denominator part
	then res.append(num/den) to get the next fraction
	num %= den change num to the next denominator part
		check if map is already contain this den part
		if it is add (), and break
		if not, put this num in map
return res.toString()*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        if(numerator==0 || denominator==0) return "0"; //return string "0"
        string ret;
        //check if the result is negative
        if((numerator>0)^(denominator>0)){
            ret += "-";
        }
        
        //change into positive
        long num = abs((long)numerator);
        long den = abs((long)denominator);
        
        //push the integer part
        ret += to_string(num/den);
        num %= den;
        if(num==0) return ret;
        ret += ".";
        
        //push the fractional part
        unordered_map<int, int> mymap;
        mymap[num] = ret.size(); // here should put the num key
        while(num!=0){
            num *= 10;
            ret += to_string(num/den);
            num %= den;
            if(mymap.find(num) != mymap.end()){
                ret.insert(mymap[num], 1, '('); // remember '(' is a char
                // cout << ret << " " << mymap[num] << endl;
                ret += ")";
                break;
            } else {
                mymap[num] = ret.size();
                // cout << num << " " << mymap[num] << " " << ret << endl;
            }
        }
        
        return ret;
    }
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        if(numerator==0 || denominator==0) return "0"; //remember string "0"
        StringBuilder ret = new StringBuilder(); //use Stringbuilder to add string
        //check the return value is positive or negative
        ret.append(((numerator>0)^(denominator>0)) ? "-" : "");
        
        //change to positive number
        long num = Math.abs((long)numerator);//remember long
        long den = Math.abs((long)denominator);
        
        //push the integral part
        ret.append(num/den);
        num %= den;
        if(num==0) return ret.toString();
        
        //push the fractional part
        ret.append(".");
        HashMap<Long, Integer> mymap = new HashMap<>(); //length should return int
        mymap.put(num, ret.length()); //remember put the size here
        while(num!=0){
            num *= 10;
            ret.append(num/den);
            num %= den;
            if(mymap.containsKey(num)){
                ret.insert(mymap.get(num), "(");
                ret.append(")");
                break;
            } else {
                mymap.put(num, ret.length());
            }
        }
        
        return ret.toString();
    }
}

File: Fraction_to_Recurring_Decimal.swift
/*
Given two integers representing the numerator（分子） and denominator（分母） of a fraction, return the fraction in string format.

If the fractional part is repeating, enclose the repeating part in parentheses.

If multiple answers are possible, return any of them.

It is guaranteed that the length of the answer string is less than 104 for all the given inputs.

 

Example 1:

Input: numerator = 1, denominator = 2
Output: "0.5"
Example 2:

Input: numerator = 2, denominator = 1
Output: "2"
Example 3:

Input: numerator = 2, denominator = 3
Output: "0.(6)"
Example 4:

Input: numerator = 4, denominator = 333
Output: "0.(012)"
Example 5:

Input: numerator = 1, denominator = 5
Output: "0.2"
 

Constraints:

-231 <= numerator, denominator <= 231 - 1
denominator != 0
*/

/*
Solution 2
optimize solution 1

we don't need extra FracVal, 
instead, only store index where we might insert "(" would be enough
*/
class Solution {
    func fractionToDecimal(_ numerator: Int, _ denominator: Int) -> String {
        if numerator == 0 { return "0" }
        
        // key is decimal part checked val
        // val is index which we start insert this key's fractorial part
        var map = [Int: Int]()
        
        var isPositive = numerator * denominator > 0
        
        var numerator = abs(numerator)
        var denominator = abs(denominator)
        
        let quotient = numerator / denominator
        var res = String(quotient)
        
        var val = numerator % denominator
        if val == 0 { 
            return isPositive ? res : "-"+res
        }
        
        if !isPositive { res = "-"+res }
        res.append(".")
                
        var frac = [Character]()
        // check decimal part
        while val != 0 {
            if map[val] != nil { break }
            
            var cur = [Character]()
            var temp = val * 10
            while temp < denominator {
                temp *= 10
                cur.append("0")
            }
            
            cur.append(String(temp / denominator).first!)
                        
            map[val] = frac.count
            
            frac.append(contentsOf: cur)
            val = temp % denominator
        }
        
        if val != 0 {
            frac.insert("(", at: map[val]!)
            frac.append(")")
        }
                
        res.append(contentsOf: frac)
        return res
    }
}

/*
Solution 1
iterative
map

1. check isPositive result or not
2. res append numeric part first
3. res append fractorial decimal part
  - use map to store current checked val

Time Complexity: O(t) - t is when numerator/denominator recurring
Space Complexity: O(t)
*/
class Solution {
    func fractionToDecimal(_ numerator: Int, _ denominator: Int) -> String {
        if numerator == 0 { return "0" }
        
        var map = [Int: FracVal]()
        
        var isPositive = numerator * denominator > 0
        
        var numerator = abs(numerator)
        var denominator = abs(denominator)
        
        let quotient = numerator / denominator
        var res = String(quotient)
        
        var val = numerator % denominator
        if val == 0 { 
            return isPositive ? res : "-"+res
        }
        
        if !isPositive { res = "-"+res }
        res.append(".")
        
        // print(val)
        
        var frac = [Character]()
        // check decimal part
        while val != 0 {
            if map[val] != nil { break }
            
            var cur = [Character]()
            var temp = val * 10
            while temp < denominator {
                temp *= 10
                cur.append("0")
            }
            
            let q = temp % denominator
            let v = temp / denominator
            cur.append(String(v).first!)
            
            // print(val, temp, q, v)
            
            map[val] = FracVal(str: cur, q: q, index: frac.count)
            
            frac.append(contentsOf: cur)
            val = q
        }
        
        if val != 0 {
            frac.insert("(", at: map[val]!.index)
            frac.append(")")
        }
        
        // print(frac)
        
        res.append(contentsOf: frac)
        return res
    }
}

struct FracVal {
    // map[4] = FracVal(str: ["0","1"], q: 67, index: )
    // 4/333 = 0.012012
    // q = 67 ((40*10)%333)
    // str = "01" ((40*10)/333)
    // 67/333 = 0.2(012012012)
    // q = 4 (670%333)
    // str = "2"
    
    let str: [Character]
    let q: Int
    
    // index where we insert this frac
    let index: Int
}